geniza/corpus/models.py

```diff
@@ -925,6 +925,65 @@ def title(self):
         """Short title for identifying the document, e.g. via search."""
         return f"{self.doctype or _('Unknown type')}: {self.shelfmark_display or '??'}"
 
+    def dating_range(self):
+        """
+        Return the start and end of the document's possible date range, as PartialDate objects,
+        including standardized document dates and inferred Datings, if any exist.
+        """
+        # it is unlikely, but technically possible, that a document could have both on-document
+        # dates and inferred datings, so find the min and max out of all of them.
+        start_date = self.start_date if self.start_date else None
+        end_date = self.end_date if self.end_date else None
+        dating_range = [start_date, end_date]
+
+        # bail out if we don't have any inferred datings
+        if not self.dating_set.exists():
+            return tuple(dating_range)
+
+        # loop through inferred datings to find min and max
+        for dating in self.dating_set.all():
+            # get start from standardized date range (formatted as "date1/date2" or "date")
+            split_date = dating.standard_date.split("/")
+            start = PartialDate(split_date[0])
+            # use numeric format to compare to current min, replace if smaller
+            start_numeric = int(start.numeric_format(mode="min"))
+            if dating_range[0] is None or start_numeric < int(
+                dating_range[0].numeric_format(mode="min")
+            ):
+                dating_range[0] = start
+            # get end from standardized date range
+            end = PartialDate(split_date[1]) if len(split_date) > 1 else start
+            # use numeric format to compare to current max, replace if larger
+            end_numeric = int(end.numeric_format(mode="max"))
+            if dating_range[1] is None or end_numeric > int(
+                dating_range[1].numeric_format(mode="max")
+            ):
+                dating_range[1] = end
+
```

Comment:
This is a bit hard to follow (and seeing it is really making me wish that my experimental `undate` library was farther along and could be used for calculations like this!). Am I understanding correctly that you're calculating the min and max across all datings and date ranges? And you have to compare with the numeric values but return as partial dates?

I'm wondering if any of this logic could be shifted to the dating class or to partial date, but I also kind of think that isn't your problem to solve and if this is working and testing it's ok. (Maybe you can help me figure out how to translate to use cases for that `undate` should support.) My guess is that if we could implement gt and lt comparison methods on partial dates you could aggregate start and end dates and then just apply `min` and `max` - but there's probably some complexity I'm missing, and I don't think you need to implement this.

Reply:
> Am I understanding correctly that you're calculating the min and max across all datings and date ranges? And you have to compare with the numeric values but return as partial dates?

Yes! Added some comments to clarify.

> (Maybe you can help me figure out how to translate to use cases for that `undate` should support.) My guess is that if we could implement gt and lt comparison methods on partial dates you could aggregate start and end dates and then just apply `min` and `max`

I think you're right, gt and lt between PartialDates would help—then we would just have to convert the string date range to PartialDates and do the comparison directly, instead of all the int casting and `numeric_format` calls.